l10n_nl_postcodeapi/models/res_partner.py

```diff
@@ -0,0 +1,111 @@
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    @ormcache(skiparg=2)
+    def get_provider_obj(self):
+        self._cr.execute("""
+                SELECT value FROM ir_config_parameter
+                WHERE key = %s""", ('l10n_nl_postcodeapi.apikey',))
+        row = self._cr.fetchone()
+        if row and row[0] != 'Your API key' and row[0].strip():
+            from pyPostcode import Api
+            provider = Api(row[0].strip())
+            test = provider.getaddress('1053NJ', '334T')
+            if not test or not test._data:
+                raise exceptions.Warning(
+                    _('Error'),
+                    _('Could not verify the connection with the address '
+                      'lookup service (if you want to get rid of this '
+                      'message, please rename or delete the system parameter '
+                      '\'l10n_nl_postcodeapi.apikey\').'))
+            return provider
+        return False
+
+    @ormcache(skiparg=1)
+    def get_province(self, province):
+        """ Return the province or empty recordset """
+        if not province:
+            return self.env['res.country.state']
+        res = self.env['res.country.state'].search([('name', '=', province)])
+        return res[0] if res else res
+
+    @api.onchange('zip', 'street_number', 'country_id')
+    def on_change_zip_street_number(self):
+        """
+        Normalize the zip code, check on the partner's country and
+        if all is well, request address autocompletion data.
+
+        NB. postal_code is named 'zip' in OpenERP, but is this a reserved
+        keyword in Python
+        """
+        postal_code = self.zip and self.zip.replace(' ', '')
+        if (not (postal_code and self.street_number)
+                or (self.country_id and
+                    self.country_id != self.env.ref('base.nl'))):
+            return {}
+
+        provider_obj = self.get_provider_obj()
+        if not provider_obj:
+            return {}
+        pc_info = provider_obj.getaddress(postal_code, self.street_number)
+        if not pc_info or not pc_info._data:
+            return {}
+        self.street_name = pc_info._data['street']
+        self.city = pc_info._data['town']
+        self.state_id = self.get_province(pc_info._province)
+
+    @api.model
+    def fields_view_get(
+            self, view_id=None, view_type='form',
+            toolbar=False, submenu=False):
+        """ Address fields can be all over the place due to module
+        interaction. For improved compatibility add the onchange method here,
+        not in a view."""
+        res = super(ResPartner, self).fields_view_get(
+            view_id=view_id, view_type=view_type,
+            toolbar=toolbar, submenu=submenu)
+        if view_type != 'form':
+            return res
+
+        def inject_onchange(arch):
+            arch = etree.fromstring(arch)
+            for field in ['zip', 'street_number']:
```

Comment:
Because https://github.com/odoo/odoo/blob/8.0/openerp/addons/base/res/res_partner.py#L68 does a hard replace of these fields, ignoring the on_change.